### PR TITLE
ENH: add computing a selected number m of extreme eigenpairs in LaplacianNd class in _special_sparse_arrays.py

### LINKED ISSUE #19056: ENH: add computing a selected number `m` of extreme eigenpairs in `LaplacianNd `class
### Is your feature request related to a problem? Please describe.

https://github.com/scipy/scipy/pull/18963 has just created a new module `scipy/sparse/linalg/_special_sparse_arrays.py` adding `LaplacianNd `class to special sparse arrays that includes
```
    Attributes
    ----------
    eigenvalues : ndarray
        Eigenvalues of the Laplacian matrix in ascending order.
```
computing at the initialization stage all eigenvalues of the Laplacian as `np.float64`.  For large matrices this computation has significant costs growing linearly with the matrix size even the user never requests the eigenvalues to be computed. Furthermore, a typical testing of eigensolvers involving the Laplacian requires only a very small fraction of the eigenvalues, the smallest (by the absolute value) or less commonly the largest ones, but also requires the corresponding eigenvectors. Both eigenvalues and eigenvectors are known analytically.

### Describe the solution you'd like.

- Remove the computation of all eigenvalues from the `LaplacianNd `class init stage as inefficient
- Allow to compute a number `m` of the extreme eigenvalues where `m>0` is a parameter instead or all if `m=0` - the default, but computed on demand only rather than in init
- Allow to additionally compute `m` eigenvectors corresponding to the `m` eigenvalues above

### Describe alternatives you've considered.

One can use the currently available all eigenvalues and just drop most of them selecting the `m` extreme ones that are of interest. 

There is no easy alternative to computing eigenvectors analytically which is not currently implemented.

### Additional context (e.g. screenshots, GIFs)

N/A

### PR DESCRIPTION
**Pull request by lobpcg**
#### Reference issue
Example: Closes gh-19056.

#### What does this implement/fix?
add computing a selected number m of extreme eigenpairs in LaplacianNd class:
- Remove the computation of all eigenvalues from the `LaplacianNd `class init stage as inefficient
- Allow to compute a number `m` of the extreme eigenvalues where `m>0` is a parameter instead, or all if `m=None` - the default
- Allow to additionally compute `m` eigenvectors corresponding to the `m` eigenvalues above

#### Additional information
See a prior discussion in https://github.com/scipy/scipy/pull/18963

### CODE REVIEW

**Comment by lobpcg** [comment]
@rkern could you please have a look if the change and the API is acceptable here. The PR is already complete for the eigenvalues, only need to add another method for the eigenvectors.

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (j-bowhay) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:136]: Here is the more compact way to check the error message. Also would prefer a more obviously wrong input as suggested below:

```suggestion
    with pytest.raises(ValueError, match="Unknown value 'robin'"):
        LaplacianNd(grid_shape=(6, ), boundary_conditions="robin")
```

**Comment by lobpcg** [comment]
I do not understand why suddenly 
ci/circleci: build_docs — Your tests failed on CircleCI ...

**Comment by ev-br** [comment]
```
/home/circleci/repo/doc/docstring of scipy.linalg._cythonized_array_utils.ishermitian:38: WARNING: py:obj reference target not found: numpy.NaN
/home/circleci/repo/doc/docstring of scipy.linalg._cythonized_array_utils.issymmetric:40: WARNING: py:obj reference target not found: numpy.NaN
```

Same error is seen on main, so can probably be ignored for this PR (or better be fixed in a separate PR)

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/_special_sparse_arrays.py:250]: All of our partial eigensolvers have you specify whether you want the largest or smallest `m>0` eigenvalues with a separate argument. I would follow `lobpcg` here and add a `largest=True` argument.

```suggestion
    def eigenvalues(self, m=None, largest=True):
        """Return the requested number of eigenvalues.
        
        Parameters
        ----------
        m : int, optional
            The positive number of eigenvalues to return. If not provided,
            then all eigenvalues will be returned.
        largest : bool, optional
            Whether to return the largest ``m`` eigenvalues (default is ``True``).
            Otherwise, return the smallest ``m`` eigenvalues.
            
        Returns
        -------
        eigenvalues : float array
            The requested eigenvalues, in ascending order.
        """
```

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:76]: Just call `lap.eigenvalues()` where you need to.

```suggestion
```

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/_special_sparse_arrays.py:213]: Since the `{boundary_conditions!r}` representation of a string value uses single quotes, these others should remain single quotes.

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/_special_sparse_arrays.py:52]: These seem unnecessary and unrelated to the rest of this PR.

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/_special_sparse_arrays.py:36]: The full call signature should be here. The following suggestion uses my suggested API from a previous comment.

```suggestion
    eigenvalues(m=None, largest=True)
```

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py:250]: @rkern I currently give up on the idea to cover both cases in this PR, and only consider m >0 the largest (smallest by absolute value since they are all negative), corresponding to smooth eigenvectors normally used for testing in PDEs

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py:213]: it was suggested to use double quotes everywhere, so I have made the needed changes. Maybe {boundary_conditions!r} should be instead changed to using the double quotes?

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py:52]: 99% of those who heard of the "Laplacian" nowadays mean the graph Laplacian. After being asked about the distinction many times, I want the connection to be clear - the actual reason for this example that I thus would like to keep.

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: Done otherwise

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/tests/test_special_sparse_arrays.py:76]: done

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: done, without the "largest" that will be unused in this PR

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [reply scipy/sparse/linalg/_special_sparse_arrays.py:52]: I was referring to the double-quotes changes.

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/_special_sparse_arrays.py:224]:
```suggestion
        if m is None:
```

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (rkern) [reply scipy/sparse/linalg/tests/test_special_sparse_arrays.py:76]: Still looks unchanged to me. Do you need to push?

Comment (rkern) [reply scipy/sparse/linalg/tests/test_special_sparse_arrays.py:76]: Actually, I would suggest getting the all-eigenvalues array here and using that array down below where we used to instead of calling `lap.eigenvalues()` each time. Then, in the "test every ``m > 0``" clause that you added, you can call `lap.eigenvalues(m)` repeatedly.

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [reply scipy/sparse/linalg/_special_sparse_arrays.py:212]: No, it needs to remain as `!r`, which will use single-quotes when given a string. It's a Python thing not in our control.

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/tests/test_special_sparse_arrays.py:76]: Sorry, lost my focus... I'll let the current CI finish first

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py:212]: `f"{boundary_conditions!r}".replace("'", "\"")` ?

Comment (rkern) [reply scipy/sparse/linalg/_special_sparse_arrays.py:212]: No.

Comment (rkern) [reply scipy/sparse/linalg/_special_sparse_arrays.py:212]: I would revert all of the double-quotes changes you made in this PR, and then apply the one quote-consistency change you were actually asked to make, which were the `@pytest.mark.parametrize("bc", ['neumann', 'dirichlet', 'periodic'])` lines.

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:85]:
```suggestion
        for m in np.arange(1, n + 1):
```

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (rkern) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:86]:
```suggestion
            assert_array_equal(lap.eigenvalues(m), eigenvalues[-m: ])
```

**Comment by lobpcg** [comment]
@h-vetinari @rkern - I did my best guessing single vs double quote use. Could you please have a look and advise if more changes are needed? There hopefully should not be too many now.

**Comment by lobpcg** [comment]
@ilayn I produced a working code here to compute the m eigenvalues selected from m*N looped with m selected in each dimension and put it into a private function that also generates the index of the sorting to be used for selecting the eigenvectors. But I ran out of steam to write the method for m*N  eigenvectors, and could not even decide on a good format how to represent/store them such that then could be then selected using the index for the sorted eigenvalues. 

The 1D eigenvectors; see https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors_of_the_second_derivative#The_discrete_case, are given by
```
if self.boundary_conditions == 'dirichlet':
    i = np.pi * (np.arange(n) + 1) / (n + 1)
    Lev = np.sqrt(2. / (n + 1.)) * np.sin(i * (j + 1))
elif self.boundary_conditions == 'neumann':
    i = np.pi * (np.arange(n) + 0.5) / n
    Lev = np.sqrt(2. / n) * np.cos(i * j)
else:  # boundary_conditions == 'periodic'
    if j == 0:
        Lev = np.sqrt(2. / n ) * np.ones(n)
    elif j + 1 == n and n % 2 == 0:
        Lev = np.sqrt(2. / n ) * np.tile([1, -1], n//2)
    else:
        i = 2. * np.pi * (np.arange(n) + 0.5) / n
        Lev = np.sqrt(2. / n) * np.cos(i * np.floor((j + 1) / 2))
```
where `n` is the number of the grid points and the index` j in np.arange(n)`. I have not tested it, but the code looks plausible. 

If you happen to have interest and bandwidth to figure out how to write the method for the eigenvectors here that would be a great help.

**Comment by lobpcg** [comment]
All done and ready for the final review please.

Computing the eigenvectors involves using lists to quickly get a short readable code. Eigenvectors are computed one-by-one to follow the ordered eigenvalues, including the case of all which is not efficient. Thus, some optimization could be done, e.g., writing a separate fully vectorized code to compute all eigenvectors and reorder them after that. But I am happy with the current version and would prefer to take any optimization that is not trivial outside of this PR.

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: Done

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: Done

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: Done

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: Done

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (j-bowhay) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:99]:
```suggestion
            r = lap.toarray() @ eigenvectors[:, i] - eigenvectors[:, i] * eigenvalues[i]
```

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (j-bowhay) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:86]:
```suggestion
            assert_array_equal(lap.eigenvalues(m), eigenvalues[-m:])
```

**Review by j-bowhay — APPROVED** [review]
LGTM modulo two minor style nits. Will leave a week before merging in case anyone else wants to look

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py:328]: Due to symmetry, the exact values are often exact zeros, but the computed values are not (and it looks bad in docstring examples).  In this calculation, there cannot be small nonzero values other than from the round-off errors. This line corrects the round-off errors.

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (lobpcg) [reply scipy/sparse/linalg/tests/test_special_sparse_arrays.py:96]: np.longdouble ?

Comment (lobpcg) [reply scipy/sparse/linalg/tests/test_special_sparse_arrays.py:96]: Sorry, it's the same so redundant. Will remove

File: scipy/sparse/linalg/_special_sparse_arrays.py
Comment (rkern) [reply scipy/sparse/linalg/_special_sparse_arrays.py:326]: Okay. Let's expand the explanatory comment, then. Otherwise, it may get removed later by someone who thinks this is merely cosmetic.

Comment (lobpcg) [reply scipy/sparse/linalg/_special_sparse_arrays.py]: done

File: scipy/sparse/linalg/tests/test_special_sparse_arrays.py
Comment (j-bowhay) [inline scipy/sparse/linalg/tests/test_special_sparse_arrays.py:159]: Lets put this test with the others
```suggestion
    def test_boundary_conditions_value_error(self):
        with pytest.raises(ValueError, match="Unknown value 'robin'"):
            LaplacianNd(grid_shape=(6, ), boundary_conditions='robin')
```

**Pull request merged**
